compiler/noirc_frontend/src/elaborator/expressions.rs

```diff
@@ -243,11 +243,46 @@ impl<'context> Elaborator<'context> {
         (HirExpression::Literal(HirLiteral::FmtStr(str, fmt_str_idents)), typ)
     }
 
-    fn elaborate_prefix(&mut self, prefix: PrefixExpression) -> (HirExpression, Type) {
+    fn elaborate_prefix(&mut self, prefix: PrefixExpression) -> (ExprId, Type) {
         let span = prefix.rhs.span;
         let (rhs, rhs_type) = self.elaborate_expression(prefix.rhs);
-        let ret_type = self.type_check_prefix_operand(&prefix.operator, &rhs_type, span);
-        (HirExpression::Prefix(HirPrefixExpression { operator: prefix.operator, rhs }), ret_type)
+        let trait_id = self.interner.get_prefix_operator_trait_method(&prefix.operator);
+
+        let operator = prefix.operator;
+        let expr =
+            HirExpression::Prefix(HirPrefixExpression { operator, rhs, trait_method_id: trait_id });
+        let expr_id = self.interner.push_expr(expr);
+        self.interner.push_expr_location(expr_id, span, self.file);
+
+        let typ = match trait_id {
+            Some(trait_id) => {
+                match Self::prefix_operand_type_rules(&operator, &rhs_type, span) {
+                    Ok((typ, use_impl)) => {
+                        if use_impl {
+                            // Delay checking the trait constraint until the end of the function.
+                            // Checking it now could bind an unbound type variable to any type
+                            // that implements the trait.
+                            let constraint = TraitConstraint {
+                                typ: rhs_type.clone(),
+                                trait_id: trait_id.trait_id,
+                                trait_generics: Vec::new(),
+                            };
+                            self.push_trait_constraint(constraint, expr_id);
+                            self.type_check_operator_method(expr_id, trait_id, &rhs_type, span);
+                        }
+                        typ
+                    }
+                    Err(error) => {
+                        self.push_err(error);
+                        Type::Error
+                    }
```

Comment:
Looks like this chunk of code is the same as for Infix operators - can you refactor the two cases into  one helper?